test/utility/thread.c

```diff
@@ -0,0 +1,194 @@
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "test_util.h"
+
+/*
+ * A thread dedicated to appending records into a table. Works with fixed
+ * length column stores and variable length column stores.
+ * One thread (the first thread created by an application) checks for a
+ * terminating condition after each insert.
+ */
+void *
+thread_append(void *arg)
+{
+	TEST_OPTS *opts;
+	WT_CONNECTION *conn;
+	WT_SESSION *session;
+	WT_CURSOR *cursor;
+	uint64_t id, recno;
+	char buf[64];
+
+	opts = (TEST_OPTS *)arg;
+	conn = opts->conn;
+
+	id = __wt_atomic_fetch_addv64(&opts->next_threadid, 1);
+	testutil_check(conn->open_session(conn, NULL, NULL, &session));
+	testutil_check(
+	    session->open_cursor(session, opts->uri, NULL, "append", &cursor));
+
+	buf[0] = '\2';
+	for (recno = 1; opts->running; ++recno) {
+		if (opts->table_type == TABLE_FIX)
+			cursor->set_value(cursor, buf[0]);
+		else {
+			snprintf(buf, sizeof(buf),
+			    "%" PRIu64 " VALUE ------", recno);
+			cursor->set_value(cursor, buf);
+		}
+		testutil_check(cursor->insert(cursor));
+		if (id == 0) {
+			testutil_check(
+			    cursor->get_key(cursor, &opts->max_inserted_id));
+			if (opts->max_inserted_id >= opts->nrecords)
+				opts->running = false;
+		}
+	}
+
+	return (NULL);
+}
+
+/*
+ * Append into a row store table.
+ */
+void *
+thread_insert_append(void *arg)
+{
+	TEST_OPTS *opts;
+	WT_CONNECTION *conn;
+	WT_SESSION *session;
+	WT_CURSOR *cursor;
+	uint64_t i;
+	char kbuf[64];
+
+	opts = (TEST_OPTS *)arg;
+	conn = opts->conn;
+
+	testutil_check(conn->open_session(conn, NULL, NULL, &session));
+	testutil_check(session->open_cursor(
+	    session, opts->uri, NULL, NULL, &cursor));
+
+	for (i = 0; i < opts->nrecords; ++i) {
+		snprintf(kbuf, sizeof(kbuf), "%010d KEY------", (int)i);
+		cursor->set_key(cursor, kbuf);
+		cursor->set_value(cursor, "========== VALUE =======");
+		testutil_check(cursor->insert(cursor));
+		if (i % 100000 == 0) {
+			printf("insert: %d\r", (int)i);
+			fflush(stdout);
+		}
+	}
+	printf("\n");
+
+	opts->running = false;
+
+	return (NULL);
+}
+
+/*
+ * Append to a table in a "racy" fashion - that is attempt to insert the
+ * same record another thread is likely to also be inserting.
+ */
+void *
+thread_insert_race(void *arg)
```

Comment:
Is this too specialized for the utility directory?  It is only used in one test and feels like it should be part of that test (at least until we have a second use for it).